File: paleo_workbench/ui/pages/well_detail_panel.py
"""Well detail panel — the V11 per-well data view surface.

Rendered from :class:`~paleo_workbench.catalog.entity_views.WellDataView`
(the read facade over project domain + catalog). The panel owns no state
beyond the last view it was handed — every refresh re-assembles from the
authorities, so there is no second data store to drift.

Shows: identity header, per-role slots (primary badge, member count,
current version), missing roles (data-completeness at a glance), stale
downstream products, uncommitted edits, and missing sources.
"""

from __future__ import annotations

from typing import Any

from PySide6.QtCore import Qt
from PySide6.QtWidgets import (
    QFrame,
    QHBoxLayout,
    QHeaderView,
    QLabel,
    QTableWidget,
    QTableWidgetItem,
    QVBoxLayout,
    QWidget,
)

from paleo_workbench.ui import tokens


class _SectionCard(QFrame):
    def __init__(self, title: str, parent=None):
        super().__init__(parent)
        self.setObjectName("WellDetailCard")
        palette = tokens.palette_for("light")
        self.setStyleSheet(
            f"QFrame#WellDetailCard {{ background: {palette['BG_SEARCH']};"
            f" border: 1px solid {palette['BORDER']};"
            f" border-radius: {tokens.RADIUS_CARD}px; }}"
        )
        self._layout = QVBoxLayout(self)
        self._layout.setContentsMargins(
            tokens.SPACE_4, tokens.SPACE_3, tokens.SPACE_4, tokens.SPACE_4
        )
        self._layout.setSpacing(tokens.SPACE_2)
        title_label = QLabel(title)
        title_label.setStyleSheet("font-weight: 600;")
        self._layout.addWidget(title_label)
        self.body = QVBoxLayout()
        self._layout.addLayout(self.body)


class WellDetailPanel(QWidget):
    """Center-stack page showing one well's full data view."""

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("WellDetailPanel")
        self._view: Any = None
        layout = QVBoxLayout(self)
        layout.setContentsMargins(
            tokens.SPACE_4, tokens.SPACE_4, tokens.SPACE_4, tokens.SPACE_4
        )
        layout.setSpacing(tokens.SPACE_3)

        self._title = QLabel("井数据视图")
        self._title.setStyleSheet("font-size: 16px; font-weight: 600;")
        self._subtitle = QLabel("")
        self._subtitle.setStyleSheet("color: #666;")
        layout.addWidget(self._title)
        layout.addWidget(self._subtitle)

        self._roles_table = QTableWidget(0, 5)
        self._roles_table.setHorizontalHeaderLabels(
            ["角色", "资产", "主用", "当前版本", "版本数"]
        )
        self._roles_table.horizontalHeader().setSectionResizeMode(
            1, QHeaderView.ResizeMode.Stretch
        )
        self._roles_table.setEditTriggers(QTableWidget.EditTrigger.NoEditTriggers)
        self._roles_table.setSelectionBehavior(
            QTableWidget.SelectionBehavior.SelectRows
        )
        self._roles_table.verticalHeader().setVisible(False)
        layout.addWidget(self._roles_table, 2)

        self._status_row = QHBoxLayout()
        self._stale_card = _SectionCard("过期成果")
        self._edits_card = _SectionCard("未提交编辑")
        self._missing_card = _SectionCard("缺失/异常")
        for card in (self._stale_card, self._edits_card, self._missing_card):
            self._status_row.addWidget(card, 1)
        layout.addLayout(self._status_row, 1)

    # ------------------------------------------------------------------

    def set_view(self, view: Any) -> None:
        """Render a WellDataView (pass None to reset)."""
        self._view = view
        if view is None:
            self._title.setText("井数据视图")
            self._subtitle.setText("")
            self._roles_table.setRowCount(0)
            for card in (self._stale_card, self._edits_card, self._missing_card):
                self._clear_layout(card.body)
            return
        well = getattr(view, "well", None) or getattr(view, "survey", None)
        if well is None:
            return
        uwi = getattr(well, "uwi", "") or ""
        self._title.setText(well.name)
        self._subtitle.setText(f"UWI: {uwi}" if uwi else "")

        from paleo_workbench.project.roles import role_definition

        slots = [
            (role, slot)
            for role, slot in view.slots.items()
            if slot.members or slot.unresolved
        ]
        # Empty roles shown at the end as "missing" rows (data completeness).
        empty = [
            (role, slot) for role, slot in view.slots.items() if not slot.members
        ]
        self._roles_table.setRowCount(len(slots))
        for row, (role, slot) in enumerate(slots):
            display = role_definition(role).display or role
            role_item = QTableWidgetItem(display)
            role_item.setData(Qt.ItemDataRole.UserRole, role)
            self._roles_table.setItem(row, 0, role_item)
            names = "、".join(m.name for m in slot.members[:4])
            if len(slot.members) > 4:
                names += f" …(+{len(slot.members) - 4})"
            self._roles_table.setItem(row, 1, QTableWidgetItem(names))
            primary = slot.primary
            self._roles_table.setItem(
                row, 2, QTableWidgetItem("✓" if primary and primary.is_primary else "")
            )
            self._roles_table.setItem(
                row, 3,
                QTableWidgetItem(
                    (primary.current_version_id or "—") if primary else "—"
                ),
            )
            version_sum = sum(m.version_count for m in slot.members)
            self._roles_table.setItem(row, 4, QTableWidgetItem(str(version_sum)))

        self._fill_list_card(
            self._stale_card.body,
            [
                (
                    f"{item.stage} · {item.version_id}"
                    + (" · pinned" if item.pinned else "")
                )
                for item in (view.stale_items or [])[:8]
            ] or ["无过期成果"],
        )
        self._fill_list_card(
            self._edits_card.body,
            [
                f"{edit.state} · {edit.source_version_id}"
                for edit in (view.uncommitted_edits or [])[:8]
            ] or ["无未提交编辑"],
        )
        missing_lines = [f"角色缺失: {role}" for role, _ in empty if role != "other"]
        missing_lines += [
            f"源文件缺失: {asset_id}"
            for asset_id in (view.missing_source_asset_ids or [])[:4]
        ]
        self._fill_list_card(self._missing_card.body, missing_lines or ["—"])

    @staticmethod
    def _clear_layout(layout) -> None:
        while layout.count():
            item = layout.takeAt(0)
            widget = item.widget()
            if widget is not None:
                widget.deleteLater()

    def _fill_list_card(self, layout, lines: list[str]) -> None:
        self._clear_layout(layout)
        for line in lines:
            label = QLabel(line)
            label.setWordWrap(True)
            layout.addWidget(label)

File: libs/ui_wellseis/src/qt/well_detail_panel.cpp
#include <pwb/ui_wellseis/qt/well_detail_panel.hpp>

#include <QFrame>
#include <QHBoxLayout>
#include <QHeaderView>
#include <QLabel>
#include <QPushButton>
#include <QTableView>
#include <QVBoxLayout>

#include <pwb/ui_wellseis/qt/object_table_model.hpp>
#include <pwb/ui_wellseis/well_detail.hpp>

namespace pwb::ui_wellseis::qt {

namespace {

QString qs(const std::string& text) {
    return QString::fromUtf8(text.data(), static_cast<int>(text.size()));
}

QFrame* section_card(const QString& title_text, QVBoxLayout** body_out,
                     QWidget* parent) {
    auto* card = new QFrame(parent);
    card->setObjectName(QStringLiteral("WellDetailCard"));
    card->setStyleSheet(QStringLiteral(
        "QFrame#WellDetailCard { background: #f6f7f9; border: 1px solid "
        "#d8dbe0; border-radius: 6px; }"));
    auto* card_layout = new QVBoxLayout(card);
    card_layout->setContentsMargins(12, 8, 12, 12);
    card_layout->setSpacing(4);
    auto* title = new QLabel(title_text, card);
    QFont title_font = title->font();
    title_font.setBold(true);
    title->setFont(title_font);
    card_layout->addWidget(title);
    auto* body = new QVBoxLayout();
    card_layout->addLayout(body);
    *body_out = body;
    return card;
}

}  // namespace

WellDetailPanel::WellDetailPanel(QWidget* parent) : QWidget(parent) {
    setObjectName(QStringLiteral("WellDetailPanel"));
    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(12, 12, 12, 12);
    layout->setSpacing(8);

    title_ = new QLabel(QStringLiteral("井数据视图"), this);
    QFont title_font = title_->font();
    title_font.setPointSizeF(title_font.pointSizeF() + 3);
    title_font.setBold(true);
    title_->setFont(title_font);
    subtitle_ = new QLabel(this);
    subtitle_->setStyleSheet(QStringLiteral("color: #6b6f76;"));
    layout->addWidget(title_);
    layout->addWidget(subtitle_);

    roles_table_ = new QTableView(this);
    roles_model_ = new StringTableModel(this);
    roles_model_->set_columns({
        {QStringLiteral("role"), QStringLiteral("角色")},
        {QStringLiteral("members"), QStringLiteral("资产")},
        {QStringLiteral("primary"), QStringLiteral("主用")},
        {QStringLiteral("current"), QStringLiteral("当前版本")},
        {QStringLiteral("versions"), QStringLiteral("版本数")},
    });
    roles_table_->setModel(roles_model_);
    roles_table_->horizontalHeader()->setSectionResizeMode(
        1, QHeaderView::Stretch);
    roles_table_->setEditTriggers(QAbstractItemView::NoEditTriggers);
    roles_table_->setSelectionBehavior(QAbstractItemView::SelectRows);
    roles_table_->verticalHeader()->setVisible(false);
    layout->addWidget(roles_table_, 2);

    auto* status_row = new QHBoxLayout();
    status_row->addWidget(
        section_card(QStringLiteral("过期成果"), &stale_body_, this), 1);
    status_row->addWidget(
        section_card(QStringLiteral("未提交编辑"), &edits_body_, this), 1);
    status_row->addWidget(
        section_card(QStringLiteral("缺失/异常"), &missing_body_, this), 1);
    layout->addLayout(status_row, 1);

    auto* close_btn =
        new QPushButton(QStringLiteral("← 返回资产列表"), this);
    connect(close_btn, &QPushButton::clicked, this,
            &WellDetailPanel::close_requested);
    layout->addWidget(close_btn, 0, Qt::AlignLeft);
}

void WellDetailPanel::fill_card(QVBoxLayout* body,
                                const std::vector<std::string>& lines) {
    while (QLayoutItem* item = body->takeAt(0)) {
        delete item->widget();
        delete item;
    }
    for (const std::string& line : lines) {
        auto* label = new QLabel(qs(line));
        label->setWordWrap(true);
        body->addWidget(label);
    }
}

void WellDetailPanel::set_view(const WellDataViewSlice& view, bool has_view) {
    has_view_ = has_view;
    view_ = view;
    if (!has_view) {
        clear();
        return;
    }
    title_->setText(qs(well_detail_title(view)));
    subtitle_->setText(qs(well_detail_subtitle(view)));

    const std::vector<WellRoleRow> rows = well_role_rows(view);
    std::vector<std::string> keys;
    std::vector<std::vector<QString>> cells;
    keys.reserve(rows.size());
    cells.reserve(rows.size());
    for (const WellRoleRow& row : rows) {
        keys.push_back(row.role);
        cells.push_back({qs(row.role_display), qs(row.member_names),
                         qs(row.primary_mark), qs(row.current_version_text),
                         QString::number(row.version_count_sum)});
    }
    roles_model_->set_rows(keys, cells);

    fill_card(stale_body_, well_stale_lines(view, /*truncate_ids=*/false));
    fill_card(edits_body_, well_edit_lines(view));
    fill_card(missing_body_, well_missing_lines(view));
}

void WellDetailPanel::update_stale(
    const std::vector<StaleItemSlice>& items) {
    if (!has_view_) {
        return;
    }
    view_.stale_items = items;
    fill_card(stale_body_, well_stale_lines(view_, /*truncate_ids=*/true));
}

void WellDetailPanel::clear() {
    has_view_ = false;
    view_ = WellDataViewSlice{};
    title_->setText(QStringLiteral("井数据视图"));
    subtitle_->setText(QString());
    roles_model_->set_rows({}, {});
    for (QVBoxLayout* body : {stale_body_, edits_body_, missing_body_}) {
        while (QLayoutItem* item = body->takeAt(0)) {
            delete item->widget();
            delete item;
        }
    }
}

}  // namespace pwb::ui_wellseis::qt
